registry/storage/catalog.go

```diff
@@ -81,78 +79,95 @@ func (reg *registry) Remove(ctx context.Context, name reference.Named) error {
 	return reg.driver.Delete(ctx, repoDir)
 }
 
-// lessPath returns true if one path a is less than path b.
-//
-// A component-wise comparison is done, rather than the lexical comparison of
-// strings.
-func lessPath(a, b string) bool {
-	// we provide this behavior by making separator always sort first.
-	return compareReplaceInline(a, b, '/', '\x00') < 0
+// getRepositories is a helper function for getRepositoriesRec calls
+// the function fn with a repository path, if the current path looked
+// at is a repository and is lexicographically after last. It is possible
+// to return driver.ErrSkipDir, if there is no interest in any repositories
+// under the given `repoPath`, or call ErrStopRec if the recursion should stop.
+func (reg *registry) getRepositories(ctx context.Context, root, last string, fn func(repoPath string) error) error {
+	midFn := fn
+
+	// middleware func to exclude the `last` repo
+	// only use it, if there is set a last.
+	if last != "" {
+		midFn = func(repoPath string) error {
+			if repoPath != last {
+				return fn(repoPath)
+			}
+			return nil
+		}
+	}
+
+	// call our recursive func, with the midFn and the start path
+	// of where we want to find repositories.
+	err := reg.getRepositoriesRec(ctx, root, root, last, midFn)
+	if err == ErrStopRec {
+		return nil
+	}
+	return err
 }
 
-// compareReplaceInline modifies runtime.cmpstring to replace old with new
-// during a byte-wise comparison.
-func compareReplaceInline(s1, s2 string, old, new byte) int {
-	// TODO(stevvooe): We are missing an optimization when the s1 and s2 have
-	// the exact same slice header. It will make the code unsafe but can
-	// provide some extra performance.
+// getRepositoriesRec recurse through all folders it the `lookPath`,
+// there it will try to find repositories. See getRepositories for more.
+func (reg *registry) getRepositoriesRec(ctx context.Context, root, lookPath, last string, fn func(repoPath string) error) error {
```

Comment:
Can we rename this to `repositoriesWalk` or `walkPath` something along those lines? The `Rec` suffix "leaks" an implementation detail -- I understand this is an unexported `func`, but having `getRepo` and `getRepoRec` can be confusing.